dynet-training/train.py

```diff
@@ -0,0 +1,214 @@
+    bilstm_repr = self.sentence_encoder.add_to_cg(sentence, train)
+    n = len(sentence)
+    stack = []
+    buffer = list(range(n))
+    losses = []
+    predicted_actions = []
+    arc_history = []
+    for gold_action_type in action_types:
+      features = extract_features(stack, buffer, bilstm_repr)
+      logits = self.action_classifier.add_to_cg(features)
+      probs = dy.softmax(logits)
+      loss = -dy.log(dy.pick(probs, gold_action_type.value))
+      losses.append(loss)
+      if train:
+        probabilities = probs.npvalue()
+      else:
+        actions_mask = get_valid_actions_mask(stack, buffer)
+        probabilities = probs.npvalue() * np.array(actions_mask)
+      predicted_action = probabilities.argmax()
+      predicted_actions.append(predicted_action)
+      #TODO: Arc history is not needed at this point
+      if train:
+        apply_action(gold_action_type, stack, buffer, arc_history)
+      else:
+        # TODO: count incomplete parses / do head accuracy.
+        predicted_action_type = ArcStandardAction(predicted_action)
+        apply_action(predicted_action_type, stack, buffer, arc_history)
+    loss = dy.esum(losses)
+    loss_value = loss.value()
+    gold_actions = [action_type.value for action_type in action_types]
+    accuracy = ArcStandardModel.compute_accuracy(gold_actions,
+                                                 predicted_actions)
+    if train:
+      loss.backward()
+      self.trainer.update()
+    return predicted_actions, loss_value, accuracy
+
+  @staticmethod
+  def compute_accuracy(gold_actions: List[int], predicted_actions: List[int]):
+    no_actions = len(gold_actions)
+    correct_predictions = 0
+    for i in range(no_actions):
+      if gold_actions[i] == predicted_actions[i]:
+        correct_predictions += 1
+    accuracy = correct_predictions/no_actions
+    return accuracy
+class SentenceEncoder():
+  
+  def __init__(self, model: dy.Model, vocab_size: int):
+    self.input_lookup = model.add_lookup_parameters((vocab_size, EMBEDDINGS_SIZE))
+    self.enc_fwd_lstm = dy.LSTMBuilder(
+      LSTM_NUM_OF_LAYERS, EMBEDDINGS_SIZE, LSTM_STATE_SIZE, model)
+    self.enc_bwd_lstm = dy.LSTMBuilder(
+      LSTM_NUM_OF_LAYERS, EMBEDDINGS_SIZE, LSTM_STATE_SIZE, model)
+    
+  def add_to_cg(self, tokens, train=False):
+    """
+    Returns:
+      a list of BiLSTM representation (a representation for each token in the
+      sentence). Each element is a vector of size BILSTM_STATE_SIZE
+    """
+    embeddings = [self.input_lookup[token] for token in tokens]
+    if train:
+      self.enc_fwd_lstm.set_dropouts(0.2, 0.2)
+      self.enc_bwd_lstm.set_dropouts(0.2, 0.2)
+    fwd_init_state = self.enc_fwd_lstm.initial_state()
+    bwd_init_state = self.enc_bwd_lstm.initial_state()
+    fw_exps = fwd_init_state.transduce(embeddings)
+    bw_exps = bwd_init_state.transduce(reversed(embeddings))
+    bilstm_repr = [dy.concatenate([f, b]) for f, b in zip(fw_exps, reversed(bw_exps))]
```

Comment:
Maybe we should have the name shortcuts for forward and backward (now, we have: `fwd_lstm` - `bwd`, `fw_exps` - `bw_exps` and `f` and `b` here)